src/compiler/checker.ts

```diff
@@ -8822,6 +8959,177 @@ module ts {
             }
         }
 
+        function checkNonThenableType(type: Type, location?: Node, message?: DiagnosticMessage) {
+            if (!allConstituentTypesHaveKind(type, TypeFlags.Any) && isTypeAssignableTo(type, getGlobalThenableType())) {

```

Comment:
A union type that includes `any` always collapses into a simple `any`. So, no need to use `allConstituentTypesHaveKind`, you can just check `TypeFlags.Any` directly on the type itself.